### PR TITLE
Dependency injection Attach command

### PR DESCRIPTION
**Pull request by Jasguerrero**
Part of: #107607

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/attach.dart [hunk 1]
```diff
@@ -58,7 +61,26 @@ import '../vmservice.dart';
...
+    required Artifacts? artifacts,
+    required Stdio stdio,
+    required Logger logger,
+    required Terminal terminal,
+    required Signals signals,
+    required Platform platform,
+    required ProcessInfo processInfo,
+    required FileSystem fileSystem,
+  }): _artifacts = artifacts,
+      _hotRunnerFactory = (hotRunnerFactory == null) ? HotRunnerFactory() : hotRunnerFactory,
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/attach.dart:76]:
```suggestion
      _hotRunnerFactory = hotRunnerFactory ?? HotRunnerFactory(),
```

File: packages/flutter_tools/lib/src/commands/attach.dart [hunk 2]
```diff
@@ -316,7 +345,7 @@ known, it can be explicitly provided to attach via the command-line, e.g.
         ).asBroadcastStream();
     }
 
-    globals.terminal.usesTerminalUi = daemon == null;
+    _terminal.usesTerminalUi = daemon == null;
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/attach.dart:348]: Seems strange to me that we set this here; I would think that we would do it somewhere earlier. I'm just talking out loud though, keeping status quo LGTM for now.

**Review by christopherfujino — APPROVED** [review]
LGTM with a nit about using `??` over ternary

**Comment by christopherfujino** [comment]
Thanks for migrating this command!

File: packages/flutter_tools/test/commands.shard/hermetic/attach_test.dart [hunk 3]
```diff
@@ -438,7 +564,16 @@ void main() {
     });
 
     testUsingContext('exits when multiple devices connected', () async {
-      final AttachCommand command = AttachCommand();
+      final AttachCommand command = AttachCommand(
```
Comment (christopherfujino) [inline packages/flutter_tools/test/commands.shard/hermetic/attach_test.dart:567]: Oh btw, are we still relying on all the overrides here? I would have hoped we could use `testWithoutContext` after this refactor.

Comment (Jasguerrero) [reply packages/flutter_tools/test/commands.shard/hermetic/attach_test.dart:567]: I just tried using `testWithoutContext` but it seems like the attach command and the `BufferLogger` used in this tests still needs a context

Comment (christopherfujino) [reply packages/flutter_tools/test/commands.shard/hermetic/attach_test.dart:567]: Annoyingly `FlutterCommand` depends on globals :(

**Pull request merged**